File: school-finder-scraper/storage/supabase_client.py
from supabase import create_client
import json

class SupabaseClient:
    def __init__(self, config):
        self.url = config['supabase_url']
        self.key = config['supabase_key']
        self.client = create_client(self.url, self.key)
    
    def upsert_schools(self, schools):
        """Dodawanie lub aktualizacja szkół w bazie danych"""
        # Konwersja danych do formatu zgodnego z bazą danych
        formatted_schools = []
        for school in schools:
            formatted_school = {
                'name': school.get('name'),
                'address': school.get('address'),
                'type': school.get('type'),
                'rspo': school.get('rspo'),
                'description': school.get('description', ''),
                'principal': school.get('principal', ''),
                'contact_info': json.dumps(school.get('contact', {})),
                'location': json.dumps(school.get('location', {})),
                'google_place_id': school.get('google_place_id', ''),
                'rating': school.get('rating'),
                'ratings_count': school.get('ratings_count', 0),
                'website': school.get('website', ''),
                'photos': json.dumps(school.get('photos', [])),
                # Pola specyficzne dla uczelni wyższych
                'founding_date': school.get('founding_date'),
                'students_count': school.get('students_count'),
                'faculties': json.dumps(school.get('faculties', [])),
                'courses': json.dumps(school.get('courses', [])),
                'last_updated': 'NOW()'
            }
            formatted_schools.append(formatted_school)
        
        # Upsert do tabeli schools
        result = self.client.table('schools').upsert(
            formatted_schools, 
            on_conflict='rspo'  # Klucz, po którym identyfikujemy duplikaty
        ).execute()
        
        return result
    
    def get_existing_schools(self):
        """Pobieranie istniejących szkół z bazy danych"""
        result = self.client.table('schools').select('rspo, name, address, last_updated').execute()
        return result.data if hasattr(result, 'data') else []
    
    def update_scraping_status(self, status, details=None):
        """Aktualizacja statusu scrapowania w bazie danych"""
        data = {
            'status': status,
            'last_run': 'NOW()',
            'details': json.dumps(details) if details else None
        }
        
        result = self.client.table('scraping_status').upsert(data).execute()
        return result

File: src/lib/scraper/storage/supabase-client.ts
/**
 * Klient Supabase dla scrapera
 */

import { createClient } from '@supabase/supabase-js';
import { Database } from '@/types/supabase';
import { ScraperConfig, ScrapingStatus } from '../index';

export class SupabaseClient {
  private client;
  
  constructor(config: ScraperConfig) {
    this.client = createClient<Database>(config.supabaseUrl, config.supabaseKey);
  }
  
  /**
   * Dodawanie lub aktualizacja szkół w bazie danych
   */
  async upsertSchools(schools: any[]): Promise<any> {
    // Konwersja danych do formatu zgodnego z bazą danych
    const formattedSchools = schools.map(school => ({
      name: school.name,
      address: school.address,
      type: school.type,
      rspo: school.rspo,
      description: school.description || '',
      principal: school.principal || '',
      contact_info: JSON.stringify(school.contact || {}),
      location: JSON.stringify(school.location || {}),
      google_place_id: school.google_place_id || '',
      rating: school.rating,
      ratings_count: school.ratings_count || 0,
      website: school.website || '',
      photos: JSON.stringify(school.photos || []),
      // Pola specyficzne dla uczelni wyższych
      founding_date: school.founding_date,
      students_count: school.students_count,
      faculties: JSON.stringify(school.faculties || []),
      courses: JSON.stringify(school.courses || []),
    }));
    
    // Upsert do tabeli schools
    const { data, error } = await this.client
      .from('schools')
      .upsert(formattedSchools, { onConflict: 'rspo' });
    
    if (error) {
      throw new Error(`Failed to upsert schools: ${error.message}`);
    }
    
    return data;
  }
  
  /**
   * Pobieranie istniejących szkół z bazy danych
   */
  async getExistingSchools(): Promise<any[]> {
    const { data, error } = await this.client
      .from('schools')
      .select('rspo, name, address, last_updated');
    
    if (error) {
      throw new Error(`Failed to get existing schools: ${error.message}`);
    }
    
    return data || [];
  }
  
  /**
   * Aktualizacja statusu scrapowania
   */
  async updateScrapingStatus(
    status: 'running' | 'completed' | 'failed',
    details?: any
  ): Promise<ScrapingStatus> {
    // Sprawdzenie, czy istnieje już wpis o statusie 'running'
    if (status === 'running') {
      const { data: runningStatus } = await this.client
        .from('scraping_status')
        .select('*')
        .eq('status', 'running')
        .limit(1);
      
      if (runningStatus && runningStatus.length > 0) {
        throw new Error('A scraping process is already running');
      }
    }
    
    // Tworzenie nowego wpisu lub aktualizacja istniejącego
    const { data, error } = await this.client
      .from('scraping_status')
      .insert({
        status,
        details: details || {}
      })
      .select()
      .single();
    
    if (error) {
      throw new Error(`Failed to update scraping status: ${error.message}`);
    }
    
    return data;
  }
  
  /**
   * Pobieranie statusu scrapowania
   */
  async getScrapingStatus(): Promise<ScrapingStatus | null> {
    const { data, error } = await this.client
      .from('scraping_status')
      .select('*')
      .order('created_at', { ascending: false })
      .limit(1);
    
    if (error) {
      throw new Error(`Failed to get scraping status: ${error.message}`);
    }
    
    return data.length > 0 ? data[0] : null;
  }
}